src/io/binary_parser.cpp

```diff
@@ -119,6 +109,13 @@ long long read_variables(string filename, dd::FactorGraph &fg) {
     fg.variables[id] =
         dd::Variable(id, type_const, is_evidence, 0, upper_bound, init_value,
                      init_value, edge_count, is_observation);
+
+    // set up the default domains for multinomial
+    for (int i = 0; i < cardinality; i++) {
+      fg.variables[id].domain.push_back(i);
+      fg.variables[id].domain_map[i] = i;
+    }
```

Comment:
Actually, we could use a [*hash consing*](https://en.wikipedia.org/wiki/Hash_consing)-like technique to share one range vector instance per cardinality if you think that's a better idea.  That approach can keep the rest of the code intact and only complicate this loading part or the constructor, with the use of a global map from cardinality to a vector pointer.